### PR TITLE
Add zip methods to get feedback

### PR DESCRIPTION
**Pull request by glesica**
Thinking about adding three new methods to `Iterable<E>`:

  - `zip` - zip two iterables of the same type together into a list of lists of length two
  - `zipTo` - zip two iterables of possibly different types, combining them with an arbitrary function
  - `zipToPairs` - same as the above but combine elements using a `Pair<L, R>` type

### CODE REVIEW

File: test/src/iterable/iterable_extensions_test.dart [hunk 1]
```diff
@@ -1,3 +1,4 @@
+import 'package:extility/src/iterable/pair.dart';
```
Comment (smaifullerton-wk) [inline test/src/iterable/iterable_extensions_test.dart]: nit sort imports 😁

File: lib/src/iterable/iterable_extensions.dart [hunk 2]
```diff
@@ -35,4 +37,72 @@ extension IterableExtensions<E> on Iterable<E> {
     }
     yield last;
   }
+
+  /// Zip the elements of the iterable together with those of the
+  /// given iterable to form a list of lists (of length two) of
+  /// elements of the same as those of the inputs..
```
Comment (smaifullerton-wk) [inline lib/src/iterable/iterable_extensions.dart]: nit extra `.` at end

Comment (smaifullerton-wk) [reply lib/src/iterable/iterable_extensions.dart]: maybe "elements at the same index of those of the inputs"?

File: lib/src/iterable/iterable_extensions.dart [hunk 3]
```diff
@@ -35,4 +37,72 @@ extension IterableExtensions<E> on Iterable<E> {
...
+  ///
+  /// ['Kate', 'Pablo'].zipToContainer(
+  ///   [28, 31],
+  ///   (name, age) => {name: age},
+  /// ); // [{'Kate': 28}, {'Pablo': 31}]
+  Iterable<C> zipTo<T, C>(
+    Iterable<T> iterable,
+    C Function(E, T) combine,
+  ) sync* {
+    final left = iterator;
```
Comment (smaifullerton-wk) [inline lib/src/iterable/iterable_extensions.dart]: whoa, I'm definitely missing something here - what is `iterator`?

File: lib/src/iterable/iterable_extensions.dart [hunk 4]
```diff
@@ -35,4 +37,72 @@ extension IterableExtensions<E> on Iterable<E> {
...
+    final right = iterable.iterator;
+
+    while (true) {
+      final leftAlive = left.moveNext();
+      final rightAlive = right.moveNext();
+      if (!leftAlive && !rightAlive) {
+        break;
+      }
+
+      yield combine(left.current, right.current);
```
Comment (smaifullerton-wk) [inline lib/src/iterable/iterable_extensions.dart]: would it come in handy to write another version of this that would allow `combine` to be async?

**Review by smaifullerton-wk — APPROVED** [review]
by the power vested in me, I hereby approve this PR

File: lib/src/iterable/iterable_extensions.dart (hunk 3 above)
Comment (glesica) [reply lib/src/iterable/iterable_extensions.dart]: `Iterable.iterator` is a getter that returns an iterator over the `Iterable`... https://api.dartlang.org/stable/2.7.0/dart-core/Iterator-class.html

I use it directly here so I can move through both `Iterable`s at the same time.

File: lib/src/iterable/iterable_extensions.dart (hunk 4 above)
Comment (glesica) [reply lib/src/iterable/iterable_extensions.dart]: Interesting idea. When would that come into play? Maybe if the `combine` function had to make network requests?

Comment (smai-f) [reply lib/src/iterable/iterable_extensions.dart]: yeah, that's what I was thinking 🤷‍♀

Comment (smai-f) [reply lib/src/iterable/iterable_extensions.dart]: whoops, on another laptop haha

File: lib/src/iterable/iterable_extensions.dart (hunk 3 above)
Comment (smai-f) [reply lib/src/iterable/iterable_extensions.dart]: oh cool, and since this is an extension method on `Iterable` you can use it directly. neat!

Comment (glesica) [reply lib/src/iterable/iterable_extensions.dart]: Yeah, extension methods are A-MAZING!

**Pull request merged**